### ISSUE TITLE
[RFC] Add Support for Event Store DB

### ISSUE THREAD

**Issue by autodidaddict** [cosmonic/concordance#23]
The current capability provider implementation, part of the Concordance platform, utilizes NATS and JetStream for all of its event streaming functionality. This RFC suggests an additional, opt-in support for [EventStore DB](https://www.eventstore.com)

## Summary
For developers who already have existing investment in EventStore or who are looking to use EventStore for their application support substrate, NATS JetStream support may not meet their needs.  

## Rationale
The rationale behind this is to give developers the choice of persistent event stores: either JetStream or EventStore.

## Implementation Plan
To help put this in perspective, this next section outlines a direct comparison of the current state of Concordance with regard to which features are supported by JetStream.

### Current
The status quo.

* **Aggregates** - Concordance uses a durable consumer on the event stream for this aggregate's topic (referred to as its stream ID in Concordance docs) to deliver events to the aggregate and updates that aggregate's state in a NATS KV bucket. Commands are delivered via durable consumer but, per [this blog post](https://cosmonic.com/blog/engineering/commands-are-not-real), we don't want to maintain that behavior going forward. Events produced by the aggregate function are then written (double-ack) to the event stream.
* **Projectors** - A durable consumer for each projector is used to pull events off of the stream. Concordance knows which events the projector is interested in from link definition configuration and will only deliver those configured. Projectors do not have Concordance-managed state, nor do they emit events or handle commands.
* **Process Managers** - Process managers are the inverse of aggregates. They receive events from a durable consumer on the event stream and their function is called with the event and the process manager state, returning a list of commands to issue. Commands are currently published to a durable stream but that will also change (see previous blog post about commands). 
* **Everything else** - Everything else is a generic event handler that consumes from a durable consumer, has no state, and cannot emit events or commands.

### RFC Proposed
This describes the high-level implementation plan suggested by this RFC.

* **Aggregates** - Concordance will stream (likely via gRPC) events from a named stream in the EventStore DB (the "stream ID" as mentioned in Concordance docs) into the aggregate. Commands will be requested of the aggregate via a configurable NATS topic. Events returned in response to commands will be written to the named stream. We'll have to look into whether we'd still use a NATS KV bucket for the aggregate and process manager state, but since NATS is still required for wasmCloud, there doesn't seem to be much reason to change this.
* **Projectors** - Concordance will pull from a named stream and dispatch events to projectors based on link definition configuration. 
* **Process Managers** - Concordance will dispatch events to them in the same way as it does to projectors and aggregates. Commands returned by the process managers will be requested of the configured topic.

To recap: the durable event stream would be _optionally_ switched from JetStream to Event Store DB. All other behavior would continue as currently implemented.

Because a large amount of the functionality is reusable across these two variants, it is the current thinking that we might be able to do this with a feature flag on the existing provider rather than creating a whole new provider. This would kill two birds with one stone, by also implementing the "make commands ephemeral" goal.

## Risks
As with all things, the biggest risks are time and sanity.

**Comment by yordis** [comment]
Semi-related to this, it is critical to figure out how the Command payload could include the Optimistic Concurrency check and how to make that compatible with the implementations:

With the Event Store DB, it would be using https://developers.eventstore.com/clients/grpc/appending-events.html#handling-concurrency while the Jetstream based would be using `Nats-Expected-Last-Subject-Sequence`

**Comment by yordis** [comment]
Is there any opportunity to introduce the Event Handler over "Projector"? The reasoning is that "Projectors" came from the specialized Event Handlers that "project" to some durable state, most likely such as PostgreSQL, but the naming wouldn't make sense when the Event Handler is just a Processor that isn't necessary for databases and state.

Also, sometimes you `may` (please ignore it for now) want to avoid the Durable Consumer because the Projector keeps track of the offset, so that you would like something like `Start-From-Offset` functionality. You want the transaction between the State Mutation and Offset Commit.

I am used to having Durable Consumers regardless, but I am just bringing it to your attention. I would ignore the subject for now, but keep it in mind.

**Comment by yordis** [comment]
> Commands are delivered via durable consumer, but we don't want to maintain that behavior going forward

One of the primary reasons I am using Nats as my Command Bus is to have Durable Command Messages for X period and potentially offload to long-term storage if I want to. Inspired by systems like Stripe as an example. Granted, maybe what I want isn't precisely Durable Consumers in Nats, as long as the message lives around until X time period. Something tells me I may be misunderstanding Jetstream.


https://cosmonic.com/blog/engineering/commands-are-not-real

**From the article:**
> The conceptual problem with this approach is that the decision that drove the creation of the command Withdraw(200) was based on the state of the world at time T1.

That is related to my previous message about the Optimistic Concurrency check, https://github.com/cosmonic/concordance/issues/23#issuecomment-1694863215 as you mentioned in the blog post.

I would push back in making assumptions about the value of having the Command around. I personally want that.

And even if you go for Ephemeral Commands, you must provide an Optimistic Concurrency Check. All you have done is to move the "Queuer" to a Client called it "a Person behind a computer screen." That Person must be aware of the fact that made a decision based on "stale/cache" data and be OK with it. Otherwise, the system design must enforce such problem.

But the fundamental problem still persists: "will my command be processed based on the state that I made the decision of"? It is not a technical one per se. Purely a business decision! 

Related to: https://github.com/commanded/commanded/issues/482

**Comment by aschrijver** [comment]
If you approach this from the perspective of interface/contract-based programming..

Concordance will offer clients an `event-sourcing` contract that abstracts NFC's related to CQRS/ES. One question in this RFC is whether an additional (e.g. `eventstore`) contract is needed that abstracts the persistence of events. And another question is whether to support EventstoreDB, the quintessential reference implementation by Greg Young, either via this `eventstore` provider contract or sorta baked in via feature flags..

>To recap: the durable event stream would be optionally switched from JetStream to Event Store DB. All other behavior would continue as currently implemented.
>
>Because a large amount of the functionality is reusable across these two variants, it is the current thinking that we might be able to do this with a feature flag on the existing provider rather than creating a whole new provider.

There are many ways that someone might persist their events. If you do a search there's multiple eventstore-like data models for Postgres to be found, for instance. Feature flags might be initially easiest, but imho not the best approach then.

However, isn't contract-based programming an important part of the wasmCloud / Cosmonic offering? Reusable lego bricks are what it's about. Create with ease, flexibility and freedoms. This methodology should be key and center then. In other words "creating a whole new provider" (resignment :disappointed:) should be "hey, we just create a new provider" (delight :partying_face:). The easiest route :)

This RFC says to me "I found my lego block breaks down into multiple smaller ones" and that's to be expected and increasing my business versatility.
